# Tasks

This notebook will hold my solutions to the Tasks Assessment for Machine Learning and Statistics 2020.
(Task 1 is the only task currently available.)

The author is Angela Carpenter

## Task 1:
**Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code should not depend on any module from the standard library or otherwise. You should research the task first and include references and a description of your algorithm.**

***

### Plan
- Some information on the square root of 2 and why this number in particular would be of interest
- How this would fit in with the machine learning and statistics module. 
    * Binary representation of floats
    * Not all numbers can be calculated using a computer. Irrational numbers such as the square root of 2 falls into this category
- Methods for calculating square roots in general
    * method by hand from school days involving making a guess to begin
    * Newtons method used for the programming and scripting problem of semester 1
- How to print out the 100 decimal places when the usual number printed is 15.
    * Using a large multiple of 2 which will print a larger whole number and then extracting the digits from this.
    


## The square root of two $\sqrt{2}$
The square root of two is the number that when multiplied by itself gives the number two.
It is a geometric number as it refers to the perfect square where each side of the square is of length 1. The length of the diagonal of a perfect square is the square root of the number 2 (Pythagoras theorem).

First to Wikipedia for any information on why this particular number is noteworthy.
According to [1][Wikipedia](https://en.wikipedia.org/wiki/Square_root_of_2)., the square root of two was probably the first number known to be irrational.
>*The square root of 2, or the one-half power of 2, written in mathematics as 
$2{\sqrt {2}}$ or $2^{{1/2}}$ is the positive algebraic number that, when multiplied by itself, equals the number 2.*

>Technically, it must be called the principal square root of 2, to distinguish it from the negative number with the same property.
[1][Wikipedia](https://en.wikipedia.org/wiki/Square_root_of_2).
[2][mathsisfun](https://www.mathsisfun.com/numbers/square-root-2-irrational.html) has a nice easy explanation of why the the square root of two is an irrational number and can never be written as a fraction.

To see what the square root of two looks like I will use the `math` module or the `numpy` package.
There is a Python function for calculating square roots in the maths module which is part of the Python Standard Library. See [math module](https://docs.python.org/3/library/math.html?highlight=math#module-math).
This task should not use any libraries but the [math.sqrt](https://docs.python.org/3/library/math.html?highlight=math#math.sqrt) function could be used to check the results of my code.
`math.sqrt(x)` calculates the square root of whatever is passed in as `x`.
This task however is to calculate the square root of the number 2 to 100 decimal places. 
The square root of a number when multiplied by itself should give the number so here I check that by multiplying the result of `math.sqrt(2)` by `math.sqrt(2)`.

The results below of squaring the square root of two returned from both the `math.sqrt` function and the `np.sqrt` function is 2.0000000000000004.  This demonstrates the problem with performing calculations on the irrational number square root of two!

In [233]:
# import math module
import math
# using the sqrt function
print(math.sqrt(2))
print(math.sqrt(2)*math.sqrt(2))
import numpy as np
print(np.sqrt(2))
print(np.sqrt(2)*np.sqrt(2))
np.sqrt(2)*np.sqrt(2)==2
# just seeing if the numpy and math square roots are the exact same
math.sqrt(2) == np.sqrt(2)
math.sqrt(2)*math.sqrt(2) == np.sqrt(2)*np.sqrt(2)

1.4142135623730951
2.0000000000000004
1.4142135623730951
2.0000000000000004


True


## Binary representation of integers and floats.

#### References used here:
- [1][Floating Point Arithmetic: Issues and Limitations](https://docs.python.org/3/tutorial/floatingpoint.html) 
 
- [2] Lecture Notes from Machine Learning and Statistics, Lecturer Dr. Ian McLoughlin
The lecture on data types of this module looked at the binary representation of integers and floats.


Floating point numbers are stored with a finite number of digits so there is a finite limit on how many decimal places they can actually store. 
- Real numbers such as $\sqrt{2}$ and $\pi$ have an infinite number of places after the decimal point. Computers cannot really store such real numbers as there is only a finite number of bits available to store the number.
- Floating point numbers are actually stored using integers in two parts, for the whole number part before the decimal point and the part of the whole after the decimal point. (Think it might be called the characteristics for the whole number and the mantissa for the part after the decimal point).

- The set of $\mathbb{N}$ Natural numbers are an infinite set but it is the smallest infinite set there is and is known as countable infinite. The set of measuring numbers are all the numbers on the number line that are used for quantifying and measuring things and there is an infinite amount of such numbers between any two discrete points on the number line.
 
- Computer programs are made up of binary strings of 1's and 0's. Binary strings can be converted into integers. Every computer program can be converted into one of the counting numbers from the set of $\mathbb{N}$ . According to Turing, all computer programs are countable which means they can be put into one-to-one correspondance with the natural numbers but Turing also realised that not all numbers can be calculated or computed.

-  A rational number can be expressed as the ratio of 2 whole numbers. It can be stored as an array of two numbers. A number is rational only if it's decimal expansion becomes periodic.

- The square root of two is an irrational number and therefore cannot be expressed in the ratio of two numbers in their most simplified form.

- The implications of this for computing is that the square root of two can never be stored. An approximation of its value can be stored though. This is the same for the irrational number $pi$. Irrational numbers are often stored as procedures which are used to calculate but not store the number.
- Caution is always required when calculation statistics or performing analytics on irrational numbers. It is possible to work theoretically or to use approximations but rounding errors could interfere will the calculations.

### Floating Point Arithmetic: Issues and Limitations
This section here is based on the tutorial in the python docs at chapter 15.
[Chapter 5 of the Python Tutorial](https://docs.python.org/3/tutorial/floatingpoint.html) looks at the issues and limitations associated with floating point arithmetic. It illustrates how the decimal fraction 0.125 and the binary fraction 0.001 have identical values:

- Floating-point numbers are represented in computer hardware as base 2 (binary) fractions
- The decimal fraction 0.125 has value $\frac{1}{10}$ + $\frac{2}{100}$ + $\frac{5}{1000}$
- The binary fraction 0.001 has value $\frac{0}{2}$ + $\frac{0}{4}$ + $\frac{1}{8}$
The only real difference here is that the decimal fraction 0.125 is written in base 10 fractional notation while the binary fraction 0.001 above is written in base 2 notation.

>Unfortunately, most decimal fractions cannot be represented exactly as binary fractions. A consequence is that, in general, the decimal floating-point numbers you enter are only approximated by the binary floating-point numbers actually stored in the machine.

(It demonstrates how the fraction 1/3 can be approximated as a base 10 fraction as  0.3 or 0.33 or 0.333. No matter how many 3's you use the result will never actually be $\frac[1}{3}$ but it will be increasingly a better approximation of $\frac[1}{3}$)

The decimal value 0.1 cannot be represented exactly as a base 2 fraction because in base 2, $\frac{1}{10}$ is the infinitely repeating fraction
0.0001100110011001100110011001100110011001100110011...
If you stop at any  finite number of bits you get an approximation. 
On most machines today floats are approximated using a binary fraction. 
- The numerator uses the first 53 bits starting with the most significant bit
- The denominator as a power of 2
The binary fraction for the decimal 0.1 is `3602879701896397 / 2 ** 55` which is close to the true value of 1/10 but is not exactly equal to the true value of 0.1.
We don't notice that this is not the true value and only an approximation because of the way values are displayed. Python (and other languages) only prints *a decimal approximation to the to the true decimal value of the binary approximation stored by the machine* 

- The true decimal value of the binary approximation stored for 0.1 is actually `0.1000000000000000055511151231257827021181583404541015625`. It is printed as 0.1 though.
- Python displays a rounded value instead of all the digits as in most situations this is enough.
- The important point is that even though what we see printed (0.01) looks like the exact value of $\frac{1}{10}$, the actual value that is stored is the nearest representable binary fraction.

- The tutorial here also illustrates how there are many different decimal numbers which actually share the same nearest approximate binary fraction. The numbers `0.1`  and `0.10000000000000001` and `0.1000000000000000055511151231257827021181583404541015625` are all approximated by `3602879701896397 / 2 ** 55` . These decimal values all have the same approximation.

- String formatting can be used to produce a limited number of significant digits.
For example `format(math.pi, '.12g')` gives 12 significant digits while `format(math.pi, '.2f')` gives 2  digits after the point.
However this is simply rounding the **display** of the true machine value.

See [The Perils of Floating Point](http://www.lahey.com/float.htm)
See [str.format](https://docs.python.org/3/library/stdtypes.html#str.format)

The errors in Python float operations come from the floating-point hardware.
>The errors in Python float operations are inherited from the floating-point hardware, and on most machines are on the order of no more than 1 part in `2**53` per operation. That’s more than adequate for most tasks, but you do need to keep in mind that it’s not decimal arithmetic and that every float operation can suffer a new rounding error.

- The [decimal module](https://docs.python.org/3/library/decimal.html#module-decimal) implements decimal arithmetic suitable for accounting applications and high-precision applications (but not for this task!).
- The [fractions module](https://docs.python.org/3/library/fractions.html#module-fractions)  implements arithmetic based on rational numbers (so the numbers like 1/3 can be represented exactly).
- NumPy and [scipy](https://scipy.org) packages are recommended for mathematical and statistical operations.
- Python also has a [float.as_integer_ratio()](https://docs.python.org/3/library/stdtypes.html#float.as_integer_ratio) method which expresses the value of a float as a fraction. This can help when you want to know the exact value as a float. As the ratio is exact, it can be used to losslessly recreate the original value.


- **Representation error** refers to the fact that most decimal fractions cannot be represented exactly as binary (base 2) fractions in Python and other languages and therefore won’t display the exact decimal number you expect.
- Most machines use IEEE-754 floating point arithmetic.
- Almost all platforms map Python floats to IEEE-754 “double precision”.
- 754 doubles contain 53 bits of precision. 
The computer strives to convert 0.1 to the closest fraction that it can of the form $J/2**N$ where J is an integer containing exactly 53 bits.


- See [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) and [IEEE Standard 754 Floating Point Numbers](https://www.geeksforgeeks.org/ieee-standard-754-floating-point-numbers/) article on <geeksforgeeks.com>.
- IEEE 754 has 3 basic components:
    * The sign of the mantissa where `0` represents a positive number while `1` represents a negative number.
    * The biased exponent
    * The Normalised Mantissa
   
#### Using `float.as_integer_ratio()`.

See [float.as_integer_ratio()](https://docs.python.org/3/library/stdtypes.html#float.as_integer_ratio) method which expresses the value of a float as a fraction. This can help when you want to know the exact value as a float. As the ratio is exact, it can be used to losslessly recreate the original value.

This function is a built-in Python function. 

In [61]:
x = np.sqrt(2)
x.as_integer_ratio()

(6369051672525773, 4503599627370496)

### Binary representation of numbers
The `bin` function can be applied to any integer to get the binary representation of an integer.
The `struct` module can be used to get the binary representation of a float.

In [25]:
import struct
f = 1/3
bin(struct.unpack('!i',struct.pack('!f', f))[0])

'0b111110101010101010101010101011'

In [29]:
f = 0.1
bin(struct.unpack('!i',struct.pack('!f', f))[0])


'0b111101110011001100110011001101'

In [9]:
from decimal import *
getcontext().prec = 28
Decimal(math.sqrt(2)) 

Decimal('1.4142135623730951454746218587388284504413604736328125')

In [10]:
#1.4142135623730951454746218587388284504413604736328125
num = 1.4142135623730951454746218587388284504413604736328125
squared =num*num
math.sqrt(squared)==num

True

### Format Strings to print more digits than the default
See the section on the bottom of the notebook where I was experimenting. To be deleted.

In [245]:
print(format(math.pi, '.55g'))
print(format(math.pi, '.55f'))
print(repr(math.pi))

3.141592653589793115997963468544185161590576171875
3.1415926535897931159979634685441851615905761718750000000
3.141592653589793


In [242]:
x = 1/10
print(format(x,'.55g'))
print(format(x,'.56g'))

0.1000000000000000055511151231257827021181583404541015625
0.1000000000000000055511151231257827021181583404541015625


- The built-in [repr](https://docs.python.org/3/library/functions.html#repr) function returns a string containing a printable representation of the object.
- See also the [eval](https://docs.python.org/3/library/functions.html#eval) built-in function.

In [231]:
print(repr(0.1))
print(repr(0.10000000000000001))
print(repr(0.1000000000000000055511151231257827021181583404541015625))
import math
print(repr(math.sqrt(2)))

0.1
0.1
0.1
1.4142135623730951


In [4]:
x = 0.1000000000000000055511151231257827021181583404541015625
print(eval(repr(x)) == x)
# 2 to the power of 55
print(2**55)
# True
print(3602879701896397/ (2**55))
print((3602879701896397/ (2**55)) == 0.1)

True
36028797018963968
0.1
True


In [128]:
print(format(10**-10,'.100g'))
print(format(10**-10,'.100f'))

1.0000000000000000364321973154977415791655470655996396089904010295867919921875e-10
0.0000000001000000000000000036432197315497741579165547065599639608990401029586791992187500000000000000


## Methods for Calculating square roots
Finding the square root of a number is the inverse of finding the square of a number. The square of a number is the number times itself. The perfect squares are the squares of the whole numbers. The square root of a number is the number that when multiplied by itself gives the number.

>*The square roots of the perfect squares (e.g., 0, 1, 4, 9, 16) are integers. In all other cases, the square roots of positive integers are irrational numbers, and hence have non-repeating decimals in their decimal representations.* [2]  <https://en.wikipedia.org/wiki/Square_root>


Getting the square root of numbers that are not perfect squares is not as straightforward.
You have to look at the square roots of perfect squares that are close to your number in question. So if you are looking at the square root of 5 you would consider the square root of 4 (which is 2) and 9 (which is 3). Therefore the square root of 5 must lie between 2 and 3. Similarly for the square root of 10, look at the square root of the perfect square 9 (which is 3) and 16 (which is 4). So the square root of 10 must lie between 3 and 4.
Looking at the square roots of the nearest perfect squares on either side of the number in question will be the starting point for finding the square root of a non-perfect square. It involves some guesswork.

As mentioned earlier, the square root of two is considered a geometric number as it refers to the perfect square where each side of the square is of length 1. The length of the diagonal of a perfect square is the square root of the number 2 (Pythagoras theorem). But the number two itself is not a perfect square.




For the programming and scripting problem set I used Newton's method to get the estimate of the square root of the number, following the tutorial provide in the Tour of Go tutorial. So I'll look there again. 
Here are some of the resources I used then which may be useful here.
- [wikipedia](https://en.wikipedia.org/wiki/Newton%27s_method)
- [math.stackexchange](https://math.stackexchange.com/questions/350740/why-does-newtons-method-work)
- [mathinsight](https://mathinsight.org/newtons_method_refresher)

Newtons Method can be used to get an estimate of the square root of a number. 
Using Newtons Method, regardless of the starting estimate, once the square of the estimate gets close enough to the actual number itself, then that is a good approximation for the square root. An initial estimate is used for the square root.

https://www.math.ubc.ca/~pwalls/math-python/roots-optimization/newton/




#### Newtons method
Here is the code using Newton's Method adapted from the tour of go that I used for the square root in the programming and scripting problem set. (problem 7)
It was adapted from the example in the Tour of Go tutorial.

In [204]:
# The code here using Newtons Method is adapted from  https://tour.golang.org/flowcomtrol/8
x= 2
guess = 1
# while the absolute difference between the guess squared and the number in question is greater than 0.01
while abs((guess * guess) - x )> 0.000000001:
        # update the guess using the formula 
    guess -=((guess * guess)-x)/(2 * guess) 
    print(guess)
    # the approximation for the square root will be the final value of estimate from the  while loop
 # assign the final estimate to sqroot   
sqroot = guess  
sqroot

1.5
1.4166666666666667
1.4142156862745099
1.4142135623746899


1.4142135623746899

### Putting Newtons code in a function. 
Here I am back looking at Newton's code but bringing the string formatting and extraction into it to get the extra digits. This is not working in the same way as the other function I used below. It also produces a different answer after the 15th decimal place. Its actually larger than the other school method I was using below rather than smaller. I could get more than 100 places with the other way. This way the computer hangs if I try using a number larger than 2 times 10 to the power of 150.

In [50]:
def newt_sqrt(Num):
    # can take the first estimate to be the number itself
    Est = Num
    # repeat this until the difference between the Estimate Squared minus the number is small enough
    while abs(Est*Est - Num) > 0.0000000000001:
        # Update the new estimate to be the current estimate minus the Number
        # Est = Est - (Est**2 - Num)/(2*Est)
        Est -= ((Est*Est)- Num)/ (2 * Est)
    # formatting the estimate to get the first 100 digits using fstrings and string extraction 
    result = f"The estimate square root is {Est}"
    print(result)
    format_Est =format(Est,'.100g')[1:90]  
    return ('1'+'.'+format_Est)

   


In [59]:
#print(newt_sqrt(9))

ans1 = newt_sqrt(2*10**150)
print(ans1)
print(len(ans1))

The estimate square root is 1.414213562373095e+75
1.414213562373095065861871032235316220382455390571766043858573365901411745792
77


### Calculate the square root of number using Babylonian method
There are several websites which show how to get the square roots of numbers in the way I would have learnt back in school. Finding the square root of a number that is not a whole number begins with some guesswork. You begin by looking at the nearest perfect squares on both sides of the number. I will work through an example here to ilustrate.

- https://en.wikipedia.org/wiki/Square_root

- https://www.homeschoolmath.net/teaching/square-root-algorithm.php

- http://www.math.com/school/subject1/lessons/S1U1L9DP.html

The post here at https://blogs.sas.com/content/iml/2016/05/18/newtons-method-babylonian-square-root.html explains simply how the algorithm thought in school for calculating the estimating a square root from an arbitrary guess is an iterative procedure. Given any positive number $S$ and any positive number $x_0$, apply the following formula:
$$x_{n+1} = \frac{(x_n + S/x_n)}{2}$$ until the iteration converges to the square root of $S$.
 

The same author has a post on the [Babylonian method for finding square roots by hand](https://blogs.sas.com/content/iml/2016/05/16/babylonian-square-roots.html). (This seems to be the way I was trying to work out below.)
Suppose you are given any positive number $S$. To find it's square root do the following:
1. Make an initial guess. This guess can be any positive number. This is $x_0$
2. Improve the guess by applying the formula $x_1 = \frac{(x_0 + \frac{S}{x_0})}{2}$.
   The number $x_1$ is a better approximation to the square root of $S$.
3. Iterate until convergence. 
   Apply the formula $x_{n+1}=\frac{(x_n+\frac{S}{x_n})}{2}$ until the process converges.

    * Convergence is achieved when the  digits of $x_{n+1}$ and $x_n$ agree to as many decimal places as you desire.

I am going to rewrite this part using the notation from above. 
    

#### Example  old school method- the Babylonian method!
Make a guess of what the root of the number in question could be based on the square root of the nearest perfect squares on either side.  Add the result of this to the number and question and divide by two to get the average as the new starting guess. Keep repeating

#### Example on square root of 10
1. Find two perfect square roots that the number in question lies between. 
    * For example if you want to find the square root of 10, you are looking at the square root of 9 which is 3 and square root of 16 which is 4.
2. Take one of those square roots as an estimate and divide the number in question by it.
    * S = 10
    * The square root of 10 must lie between 3 and 4
    * Taking 3 here as the estimate $x_0=3$
    * Divide number in question by the starting estimate S/x0 = 10/3 = 3.3333 = X1
    
 
3. Take the average of the latest estimate x1 (3.3333) and S (from the previous step above) and the Number.
    *  The average of the number 3 and the starting estimate x0 becomes the next estimate  x1

    * x1 = (x0 +S/x0)/2 = (3 + 10/3)/2 = 3.16666667
    * x2 =  (x1 + S/x1)/2 = 3.162280686800554 
    * x3 = (x2 + S/x2)/2 = (3.162280686800554 +10/3.162280686800554)/2 = 3.1622776601698277
    * x4 = (x3+ S/x3)/2 = (3.1622776601698277 + 10/3.1622776601698277)/2 = 3.162277660168379
    * x5 = (x4 + S/x4)/2 = (3.162277660168379 + 10/3.162277660168379)/2 = 3.162277660168379
    
    `math.sqrt(10)` gives 3.1622776601683795 
    
#### Example of square root of 2

1. Find two perfect square roots that the number in question lies between. 
 
2. Take one of those square roots as an estimate and divide the number in question by it.
    * S = 2
    * The square root of 2 must lie between 1 and 2
    * Taking 1 here as the estimate $x_0=1$
    * Divide number in question by the starting estimate S/x0 = 2/1 = 2 = X1
    
 
3. Take the average of the latest estimate x1 (2) and S (from the previous step above) and the Number.

    * x1 = (x0 +S/x0)/2 = (2 + 2/1)/2 = 2
    * x2 =  (x1 + S/x1)/2 = (2 + 2/2)/2 =1.5
    * x3 = (x2 + S/x2)/2 = (1.5 + 2/1.5)/2 = 1.4166666666666665
    * x4 = (x3+ S/x3)/2 = (1.4166666666666665 + 2/1.4166666666666665)/2 = 1.4142156862745097
    * x5 = (x4 + S/x4)/2 = (1.4142156862745097 + 2/1.4142156862745097)/2 = 1.4142135623746899
    
    `math.sqrt(10)` gives 3.1622776601683795 

- The square root of 2 truncuated to 50 decimal places is 1.41421356237309504880168872420969807856967187537694
See (https://en.wikipedia.org/wiki/Square_root).


In [78]:
(3.16666667 + (10/3.1666667))/2
(3.162280686800554 +10/3.162280686800554)/2
(3.1622776601698277 + 10/3.1622776601698277)/2
(3.162277660168379 + 10/3.162277660168379)/2 
(2 + 2/1)/2
(2 + 2/2)/2 
(1.5 + 2/1.5)/2
(1.4166666666666665 + 2/1.4166666666666665)/2
(1.4142156862745097 + 2/1.4142156862745097)/2

1.4142135623746899

In [68]:
import math
math.sqrt(10)

3.1622776601683795

### Applying the above to get the square root of 2.

$1^2=1$ and $2^2=4$, therefore the square root of 2 must fall between 1 and 2.
Use 1 as the starting guess.  (The square root of two is not two.) 
Putting the above algorithm into code...
Use the notation S for the number and $x_n$, $x_0$, $x_1$, $x_{n+1}$ for the estimates at each step.
Actually I am just going to use S for the number to get the square root of and `est` for the estimates at each step as I am using a while loop so the value of the estimate will be updated at each iteration, representing the $X_n$ and $x_{n+1}$ estimates

In [130]:

S = 2
est =1

# while the difference between 2 and the calculated square root is greater than a specified amount
while abs(S -(est * est)) > 0.000000000000001:
    
    est= (est + S/est)/2

print(est)

result = ('%.60f' %est)
print(result)
print(type(result))
print('%.60g' %est)

1.414213562373095
1.414213562373094923430016933707520365715026855468750000000000
<class 'str'>
1.41421356237309492343001693370752036571502685546875


In [106]:
# define the function
def mySqrtA(x):
    """
    takes  one input which is the number to calculate the square root of
    could also take the 
    
    """
    S = 2 
    est = 1
 
    while abs(S -(est * est)) > 0.000000000000001:
        est= (est + S/est)/2
        
    return(est)
        
print(mySqrtA(2))   

1.414213562373095


In [117]:
# define the function, without asking for an argument here
def mySqrt2():
    """
    Here I am going to pass in the big multiple of 2
    
    """
    S = 2
    est = 1
 
    while abs(S -(est * est)) > 0.000000000000001:
        est= (est + S/est)/2
        
    return(est)
        
print(mySqrt2())

1.414213562373095


### Rounding / Accuracy etc
Moving this up from the bottom of the notebook: probably delete later

At some stage the results gets rounded which is the main problem here! From Wikipedia the square root of 2 to 50 decimal places is `1.41421356237309504880168872420969807856967187537694`
The square root of 2 using `np.sqrt` to 50 decimal places is 1.41421356237309514547462185873882845044136047363281.
The power of two is the opposite of getting the square root.
The `**` operator can be used to get powers in Python.
This is where I started before so it should be moved up or deleted.

While the algorithm itself will calculate approximately the square root of two (accurate to 15 digits at least), it will not print out 100 decimal places so easily. The number before the decimal point cannot be rounded though so I decided to try using a large multiple of 2. In the same way that the square root of 4 is 2, square root of 400 is 20, (or square root of 9 is 3 and the square root of 900 is 30). 
- The square root of 4 times 100 (400) is the same as 100 times the sqaure root of 4.
- The square root of 900 is the same as 100 times the square root of 9.

If I get the square root of a multiple of two and then divide by that multiple I might be able to increase the number of decimal places by capturing the whole number part of a multiple of 100 and then dividing by 100.
(I am using the math module here just to see what the square root of a multiple of 2 would look like)

So first try multiplying by 100, then square root of the result and divide by 100.
Can use the `**` to raise to power of in Python. See [python docs](https://docs.python.org/3/tutorial/introduction.html#using-python-as-a-calculator)

10 to the power of 2 is 100, 10 to the power of 10 is 10000000000
10 to the power of 100 is a big number! I can get the length of this by using the length function on a string. First need to convert into a string using the `str` function and then use the `len` function on the result of that or `len(str()` in one go.

Look back again at [python docs on floating points numbers.](https://docs.python.org/3/tutorial/floatingpoint.html) which goes into detail as to why
>Unfortunately, most decimal fractions cannot be represented exactly as binary fractions. A consequence is that, in general, the decimal floating-point numbers you enter are only approximated by the binary floating-point numbers actually stored in the machine.

### Extracting the digits from the string
I now need to take away the `e+200` from the end of the string.
See [strings tutorial](https://docs.python.org/3/tutorial/introduction.html#strings) from the Python docs. I can just take the first 102 characters including the 1 and the decimal point and not including the `e+200`

### Modifying my function above
Bringing the extra steps for formatting the result into the function instead of being applied to the function after. However so far when I try this inside the function, it returns scientific notation.

In [138]:
# define the function
def sqrt2():
    """
    takes  one input which is the number to calculate the square root of
    
    """
    S = 2 # This is the number we want to get the square root of
    
    S = S*(10**200) # 2 times 10^100

    est = S
    # can set the precision here or hardcode it i
   
    # while the difference between the number and the square of the guess is greater than 0
    while abs(S -(est * est)) > 0.000000000000001:
        est= (est + S/est)/2
        
    return(est)

    print("The root of 2 times 10 to power of 200 is: ",est)
   
    format_est =format(est,'.100g')
    # extract the first 100 digits
    print(format_est)
    sqrtoftwo =format_est[0:101]
    print(sqrtoftwo)

sqrtof2 = sqrt2()

In [139]:
# from my function using babylonian method
sqrt2format =format(sqrtof2,'.100g')
sqrt2format[0:101]

'1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877'

## 

In [142]:
import math
math_root2 = math.sqrt(2*(10**200))
print(" The square root of two is \n{:.100g}".format(root2))
print(math_root2)


root2 = math.sqrt(2*(10**200))
print(" The square root of two is \n{:.100g}".format(root2))

#from my sqrt2 function
sqrt2format[0:101]


 The square root of two is 
1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100
1.414213562373095e+100
 The square root of two is 
1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100


'1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877'

Here I am looking to see if the logic I used above will give the same results as the `math` module and `numpy` would return for very large multiples of 2.
The results that both the Numpy and `math` modules produce for the square root of 2 times 10 to power of 200 is actually very close to the result I get using my function from the Babylonian method. It seems to be the same for the first 100 places. 



#### Some notes
- Instead of hardcoding in the precision amount, I could pass it as a parameter. This would help with testing it as I could change the precision to see how the results change.
- Converting from one format to another, from a float to a string might not actually be returning the exact answer. 
- My best solution below was getting the square root of a large multiple of 2 and then extracting the digits into a string. I need to incorporate this into the function, both the larger number 2 times 10 to power of 200. 
- My function is not calculating the correct number compared to Wikipedia. It is rounding down after the first 15 digits. This is the challenge really so I need to modify my function to become more accurate. At least the printing of the digits part is working!



# References
- https://www.latex-tutorial.com/tutorials/amsmath/
- https://www.math-linux.com/latex-26/faq/latex-faq/article/latex-natural-numbers

Read https://kodlogs.com/blog/715/convert-scientific-notation-to-decimal-python